pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -123,17 +154,24 @@ func NewLazyBinaryReader(
 		level.Debug(logger).Log("msg", "built index-header file", "path", path, "elapsed", time.Since(start))
 	}
 
-	return &LazyBinaryReader{
+	reader := &LazyBinaryReader{
```

Comment:
A lot of complexity and fiddlyness comes from each request having their own channel. Without being a 100% master of the below code, is the following simpler structure viable?

* LazyBinaryReader (_LBR_) maintains one chan `r.C`
* When someone performs a lookup on the LBR,
   * if there isn't an outstanding request already, we kick one off and create `r.C`. Return `r.C` to the caller.
   * otherwise, there's either a load in progress or the load finished already. return `r.C` to the caller.
* when LBR finishes loading a reader, it stores the result in an atomic `struct{Reader, error}` type thing and closes `r.C`.
* The callers (`getOrLoadReader`, I think) can use `select` to either wait on `r.C` or on their context to die. if `r.C` is signaled, they call `lbr.Result()` to obtain the (reader, error) pair.

`inUse` goes back to being a member of the LBR struct. Unloading waits on `inUse`, then unloads the file. Something like that.

Is that viable, and would a control loop even be necessary?

Reply:
I _think_ this goes against the proverb:
> _[Don't communicate by sharing memory, share memory by communicating.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=2m48s)_

I think that the atomic value and signalling is more error prone and harder to follow than a channels approach, where you ask for something, and you get it back.

Reply:
I somewhat agree with Oleg. But also would this work with unloading a loaded block - isn't there a race between taking the address of the closed channel and replacing it with a new channel?

I guess we can protect the channel with a mutex, but then we're using 3 sync primitives (is that what you call them?) to achieve the job (atomic, channel, mutex), which looks a bit convoluted.